library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package scala.scalajs.macroImpls
+
+import scala.annotation.tailrec
+import scala.scalajs.js
+
+import Compat210._
+
+/** Macros for js.use(x).as[T]
+ *
+ *  This implements a structural typechecker conforming to the JavaScript calling
+ *  convention in Scala.js' export and facade system.
+ *
+ *  @author Tobias Schlatter
+ */
+object UseAsMacros {
+  // Import macros only here, otherwise we collide with Compat210._
+  import scala.reflect.macros._
+  import blackbox.Context
+
+  def as_impl[A : c.WeakTypeTag, B <: js.Any : c.WeakTypeTag](
+      c: Context { type PrefixType = js.Using[_] }): c.Expr[B] = {
+    (new Macros[c.type](c)).as[A, B]
+  }
+
+  private class Macros[C <: Context { type PrefixType = js.Using[_] }](val c: C)
+      extends JSMembers with Compat210Component {
+
+    import c.universe._
+
+    private val JSNameAnnotation = typeOf[js.annotation.JSName].typeSymbol
+    private val JSExportAnnotation = typeOf[js.annotation.JSExport].typeSymbol
+    private val JSExportAllAnnotation = typeOf[js.annotation.JSExportAll].typeSymbol
+
+    /** Base classes that are allowed in a target type.
+     *  These are also the classes whose methods do not need to be provided.
+     */
+    private val JSObjectAncestors = typeOf[js.Object].baseClasses.toSet
+
+    def as[A : WeakTypeTag, B <: js.Any : WeakTypeTag]: Expr[B] = {
+      val trgTpe = verifyTargetType(weakTypeOf[B])
+      val srcTpe = weakTypeOf[A]
```

Comment:
Or else, use `exportedMembers(srcTpe)` if `srcTpe` is not a subtype of `js.Any`, but use `rawJSMembers(srcTpe)` if it is. And then you suddenly have a macro that can also cast between a pair of raw JS types with structural type checking (which can be useful in some cases to retroactively "implement" a super-trait, e.g., sometimes core ES classes implement DOM interfaces, because those DOM interfaces are structurally supertypes of those, but that subtyping relationship cannot be expressed with `extends`).